File: 07/07.py
lines = []
with open('input.txt', 'r') as f:
    lines = f.readlines()
    
bag_dict = {}
for line in lines:
    words = line.split(' ')
    if words[4] == "no": continue
    d = { "name": "{} {}".format(words[0],words[1]), "bags": []}
    d["bags"].append({ "name": words[5] + " " + words[6], "count": int(words[4]) })
    remaining_bags = line.split(',')[1:]
    for bag in remaining_bags:
        words = [b for b in bag.split(' ') if b != ""]
        name = "{} {}".format(words[1].strip(), words[2].strip())
        count = int(words[0].strip())
        bag_info = {"name": name, "count": count}
        d["bags"].append(bag_info)
    bag_dict[d["name"]] = d["bags"]

def rec_bag_count(bag_dict, name, inner_bag_name):
    if name not in bag_dict:
        return 0
    bags = bag_dict[name]
    count = 0
    for b in bags:
        if b["name"] == inner_bag_name:
            count += 1
        else:
            count += rec_bag_count(bag_dict, b["name"], inner_bag_name)
    return count

shiny_gold_count = 0
for name, _ in bag_dict.items():
    if rec_bag_count(bag_dict, name, "shiny gold") > 0:
        shiny_gold_count += 1
print("Part 1: ", shiny_gold_count)
def rec_inner_bag_count(bag_dict, name):
    if name not in bag_dict:
        return 0
    bags = bag_dict[name]
    count = 0
    for bag in bags:
        count += bag["count"]
        count += bag["count"] * rec_inner_bag_count(bag_dict, bag["name"])
    return count
p2 = rec_inner_bag_count(bag_dict, "shiny gold")
print("Part 2: ", p2)

File: 07/07.hs
import           Data.Either        (rights)
import qualified Data.Map.Strict    as M
import           Data.Maybe         (fromMaybe)
import           Text.Parsec
import           Text.Parsec.String (Parser)
import           Text.Read          (readMaybe)

type Bag = String

type BagDef = (Bag, [(Bag, Int)])

bagDefP :: Parser BagDef
bagDefP = do
  desc <- many1 letter
  _ <- space
  color <- many1 letter
  let topBag = unwords [desc, color]
  _ <- space
  _ <- string "bags"
  _ <- space
  _ <- string "contain"
  _ <- space
  contained <- try $ many1 noBags <|> many1 innerBagP
  return (topBag, contained)

innerBagP :: Parser (Bag, Int)
innerBagP = do
  bCount <- digit
  _ <- space
  desc <- many1 letter
  _ <- space
  color <- many1 letter
  let name = unwords [desc, color]
  _ <- space
  _ <- optional $ try $ string "bags"
  _ <- optional $ try $ string "bag"
  _ <- try (char ',' <|> char '.')
  _ <- optional $ try space
  return (name, fromMaybe (-1) (readMaybe [bCount] :: Maybe Int))

noBags :: Parser (Bag, Int)
noBags = do
  _ <- string "no other bags."
  return ("", 0)

numBagsCanContain :: Bag -> M.Map Bag [Bag] -> Int
numBagsCanContain name bagMap = length $ M.filterWithKey canCarryBag $ M.delete name bagMap
  where
    canCarryBag :: Bag -> [Bag] -> Bool
    canCarryBag b [] = b == name
    canCarryBag b bags = (b == name) || any (\ib -> canCarryBag ib (M.findWithDefault [] ib bagMap)) bags

innerBagCount :: Bag -> M.Map Bag [(Bag, Int)] -> Int
innerBagCount name bagCountMap = innerBagCount' name
  where
    innerBagCount' :: Bag -> Int
    innerBagCount' name =sum(map (\(b,c) -> c + c * innerBagCount' b) bagCount)
      where
        bagCount = M.findWithDefault [] name bagCountMap

main :: IO ()
main = do
  inputs <- lines <$> readFile "input.txt"
  let parsed = map (parse bagDefP "input.txt") inputs
  let bags  = filter (not . null . fst) $ rights parsed
  let bagMap = map fst <$> M.fromList bags
  let p1 = numBagsCanContain "shiny gold" bagMap
  putStrLn $ "Part 1: " ++ show p1

  let bagCountMap = M.fromList bags
  let p2 = innerBagCount "shiny gold" bagCountMap
  putStrLn $ "Part 2: " ++ show p2
